### PR TITLE
Bump dependencies July 2024

### PR DESCRIPTION
**Pull request by mpharrigan**
- Some minor changes due to a new mypy version
 - A bunch of type errors due to operator typing and new sympy version

### CODE REVIEW

**Comment by mpharrigan** [comment]
minimal problem that doesn't even use sympy

```python
import numbers

x:numbers.Number = 5
assert isinstance(x, numbers.Number)
print(2**x)
```

```
qualtran/op_check.py:3: error: Incompatible types in assignment (expression has type "int", variable has type "Number")  [assignment]
qualtran/op_check.py:5: error: No overload variant of "__pow__" of "int" matches argument type "Number"  [operator]
qualtran/op_check.py:5: note: Possible overload variants:
qualtran/op_check.py:5: note:     def __pow__(self, Literal[0], /) -> Literal[1]
qualtran/op_check.py:5: note:     def __pow__(self, Literal[0], None, /) -> Literal[1]
qualtran/op_check.py:5: note:     def __pow__(self, Literal[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25], None = ..., /) -> int
qualtran/op_check.py:5: note:     def __pow__(self, Literal[-1, -2, -3, -4, -5, -6, -7, -8, -9, -10, -11, -12, -13, -14, -15, -16, -17, -18, -19, -20], None = ..., /) -> float
qualtran/op_check.py:5: note:     def __pow__(self, int, None = ..., /) -> Any
qualtran/op_check.py:5: note:     def __pow__(self, int, int, /) -> int

```

File: dev_tools/requirements/deps/mypy.txt [hunk 1]
```diff
@@ -1,3 +1,8 @@
 # the mypy dependency file
 mypy~=1.9
 mypy-protobuf
+
+# Newer sympy versions cause many [operator] mypy errors
+# including ones seemingly unrelated to sympy
+# https://github.com/quantumlib/Qualtran/pull/1167
+sympy<1.13
```
Comment (tanujkhattar) [inline dev_tools/requirements/deps/mypy.txt:8]: Should we open an issue with sympy with a minimal example?

**Review by tanujkhattar — APPROVED** [review]

File: dev_tools/requirements/deps/mypy.txt (hunk 1 above)
Comment (mpharrigan) [reply dev_tools/requirements/deps/mypy.txt:8]: on the sympy repository? I'm not convinced this isn't our fault :p

**Comment by mpharrigan** [comment]
I'm going to merge but will follow up on the sympy stuff (either by investigating or opening an issue)

File: dev_tools/requirements/deps/mypy.txt (hunk 1 above)
Comment (tanujkhattar) [reply dev_tools/requirements/deps/mypy.txt:8]: Oh my bad, I didn't see any qualtran code in the minimal example so I thought sympy is running into errors unrelated to Qualtran. I now see the error messages were generated by Qualtran.

**Pull request merged**